Turning on ClusterRepair for edge hits only

cmantill opened the pull request:
#### PR description:

This PR turns ON ClusterRepair CPE for edge hits only. 
It adds an option to the configuration file so that is possible to choose which pixel sub-detectors/layers to run on. By default this option is: "PXB 2","PXB 3","PXB 4" which refers to layers 2,3 and 4 of PixelBarrel.
The PR further adds a flag to turn on/off CR for truncated clusters - this flag is false by default.

The PR follows from the discussion on the DPG meeting https://indico.cern.ch/event/787658/?showDate=all&showSession=6

The following are updated tags for autoCond:

### **Update pixel local reco conditions** in the offline GTs

### **Data**
New offline GT is
https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/106X_dataRun2_v4
Diff
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_dataRun2_v4/106X_dataRun2_v1
New offline relval GT is
https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/106X_dataRun2_relval_v2
Diff
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/106X_dataRun2_relval_v2/106X_dataRun2_relval_1

Only 2017 conditions are updated in GenErr, Lorentz angle and 1D template.
2D template changes in 2016 and 2017, but it was not consumed in the past till this PR.

- GenErr 
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/tags/SiPixelGenErrorDBObject_38T_forPR26263/SiPixelGenErrorDBObject_38T_v8_offline
- Lorentz angle 
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/tags/SiPixelLorentzAngle_38T_forPR26263/SiPixelLorentzAngle_v10_offline
- 1D template
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/tags/SiPixelTemplateDBObject_38T_forPR26263/SiPixelTemplateDBObject_38T_v13_offline
- 2D template 
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/tags/SiPixel2DTemplateDBObject_38T_v2_offline/SiPixel2DTemplateDBObject_38T_v1_offline

#### PR validation:

It has been tested locally on data - using cmsDriver step3 - (on JetHT and on a run for which a 2D template exists) and verified by printouts that is called only for the layers that are defined in the configuration file.
It has also been tested using standard runTheMatrix tests using 105X_dataRun2_Candidate_2019_04_03_13_37_01 GT and checking step3 independently for runs corresponding to each 2017 IOV.

cmantill committed:
add flags to run only on selected layers and damaged clusters

cmantill committed:
modify when recommend 2D

cmantill committed:
turning on CR

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/8938

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @cmantill (Cristina Ana Mantilla Suarez) for master.

It involves the following packages:

RecoLocalTracker/SiPixelRecHits
RecoTracker/TransientTrackingRecHit


@cmsbuild, @perrotta, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @GiacomoSguazzoni, @rovere, @VinInn, @mschrode, @dkotlins, @gpetruc, @ebrondol, @threus, @dgulhan this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmantill commented:
Standard runTheMatrix tests have been performed -> but they are failing due to a global tag error [1]
The current release I'm using is CMSSW_10_6_0_pre2 and sl7. 
We expect jobs to fail because there are no updated conditions for 2D templates but this error does not seem related. If anyone has any suggestions please let me know. 

[1] `----- Begin Fatal Exception 26-Mar-2019 18:37:15 CDT-----------------------
An exception of category 'StdException' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing ESSource: class=PoolDBESSource label='GlobalTag'
Exception Message:
A std::exception was thrown.
Connection on "frontier://(preferipfamily=0)(proxyconfigurl=http://grid-wpad/wpad.dat)(backupproxyurl=http://cmst0frontier.cern.ch:3128)(backupproxyurl=http://cmst0frontier1.cern.ch:3128)(backupproxyurl=\
http://cmst0frontier2.cern.ch:3128)(backupproxyurl=http://cmsbpfrontier.cern.ch:3128)(backupproxyurl=http://cmsbpfrontier1.cern.ch:3128)(backupproxyurl=http://cmsbpfrontier2.cern.ch:3128)(backupproxyurl=\
http://cmsbproxy.fnal.gov:3128)(serverurl=http://cmsfrontier.cern.ch:8000/FrontierProd)(serverurl=http://cmsfrontier1.cern.ch:8000/FrontierProd)(serverurl=http://cmsfrontier2.cern.ch:8000/FrontierProd)(s\
erverurl=http://cmsfrontier3.cern.ch:8000/FrontierProd)(serverurl=http://cmsfrontier4.cern.ch:8000/FrontierProd)/CMS_CONDITIONS" cannot be established ( CORAL : "ConnectionPool::getSessionFromNewConnecti\
on" from "CORAL/Services/ConnectionService" )
----- End Fatal Exception -------------------------------------------------
`

cmantill commented:
Quick update: I repeated the runTheMatrix standard tests on lxplus and most of them were completed (it seems that the error from before was something regarding the connection in the fermilab cluster). 

The following workflows fail:
136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM
140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18
136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017

[2] 
Exception Message:
SiPixelTemplate2D::interpolate can't find needed template ID = 6534, Are you using the correct global tag?

tvami commented:
Hi @tocheng, the PR fails because it wants to use the Prompt GTs. PR26281 seems to fix this. Can we do some kind of workaround to pass these tests?

cmantill commented:
@tvami I tried your GT: 105X_dataRun2_v8 and this works for all 3 workflows that were failing before.

But I tried the PR26281 and it still does not work, e.g for workflow 136.788:
```
No HLT prescale table found
Using default empty table with all prescales 1
%MSG
----- Begin Fatal Exception 29-Mar-2019 03:20:27 CET-----------------------
An exception of category 'DataCorrupt' occurred while
   [0] Processing  Event run: 297557 lumi: 119 event: 189755411 stream: 0
   [1] Running path 'Flag_METFilters'
   [2] Prefetching for module BooleanFlagFilter/'HBHENoiseFilter'
   [3] Prefetching for module HBHENoiseFilterResultProducer/'HBHENoiseFilterResultProducer'
   [4] Prefetching for module HcalNoiseInfoProducer/'hcalnoise'
   [5] Prefetching for module FastjetJetProducer/'ak4PFJets'
   [6] Prefetching for module PFLinker/'particleFlow'
   [7] Prefetching for module PFProducer/'particleFlowTmp'
   [8] Prefetching for module PFBlockProducer/'particleFlowBlock'
   [9] Prefetching for module PFElecTkProducer/'pfTrackElec'
   [10] Prefetching for module GsfTrackProducer/'electronGsfTracks'
   [11] Prefetching for module MeasurementTrackerEventProducer/'MeasurementTrackerEvent'
   [12] Prefetching for module JetCoreClusterSplitter/'siPixelClusters'
   [13] Prefetching for module PrimaryVertexProducer/'firstStepPrimaryVerticesPreSplitting'
   [14] Calling method for module TrackProducer/'initialStepTracksPreSplitting'
Exception Message:
SiPixelTemplate2D::interpolate can't find needed template ID = 715, Are you using the correct global tag?
----- End Fatal Exception -------------------------------------------------
%MSG-w BeamFitter:  AlcaBeamMonitor:AlcaBeamMonitor@endLumi  29-Mar-2019 03:20:27 CET Run: 297557 Lumi: 119
No event read! No Fitting!
%MSG
```

tocheng commented:
@tvami @cmantill 
I suspect the 2D template works for 2017.
If you look at the tag,
https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/SiPixel2DTemplateDBObject_38T_v1_offline
the first IOV covers from run 1 to run 312209 where 2016 and 2017 runs all fall into this IOV.

mmusich commented:
> @tvami @cmantill
> I suspect the 2D template works for 2017.
> If you look at the tag,
> https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/SiPixel2DTemplateDBObject_38T_v1_offline
> the first IOV covers from run 1 to run 312209 where 2016 and 2017 runs all fall into this IOV.

@tocheng I don't believe that's the problem for issue reported in https://github.com/cms-sw/cmssw/pull/26263#issuecomment-477256574, because all the other workflows run fine excepted there 3:

``` 
136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM
140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18
136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017
```
By the way,  ***phase-0  detector reconstruction is not touched intentionally by this PR***, see : 
https://github.com/cms-sw/cmssw/pull/26263/files#diff-f952152458b1577d1a7da17a1db86daaR16

As far as I can see all of the three failing workflows use the same 2D template tag: [SiPixel2DTemplateDBObject_38T_v1_prompt](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/SiPixel2DTemplateDBObject_38T_v1_prompt):

136.788 => GT: `auto:run2_data_promptlike` ([105X_dataRun2_PromptLike_v8](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/105X_dataRun2_PromptLike_v8) in[ master branch](https://github.com/cms-sw/cmssw/blob/e97a97f1206006b2fddd41fd30aa60e6d25a94db/Configuration/AlCa/python/autoCond.py#L35))
136.85   => GT: `auto:run2_data_promptlike` ([105X_dataRun2_PromptLike_v8](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/105X_dataRun2_PromptLike_v8) in [master branch](https://github.com/cms-sw/cmssw/blob/e97a97f1206006b2fddd41fd30aa60e6d25a94db/Configuration/AlCa/python/autoCond.py#L35))
140.56   => GT: `auto:run2_data_promptlike_hi` ([105X_dataRun2_PromptLike_HI_v3](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/gts/105X_dataRun2_PromptLike_HI_v3)  in [master branch](https://github.com/cms-sw/cmssw/blob/e97a97f1206006b2fddd41fd30aa60e6d25a94db/Configuration/AlCa/python/autoCond.py#L36))

and all of these GTs use the same tag [0]
As the various failing workflows fail for runs: 297557, 315489, 326479 it looks like all the 3 payloads in that tag are problematic.

```
Since: Run   Insertion Time       Payload                                   Object Type               
-----------  -------------------  ----------------------------------------  ------------------------- 
1            2018-06-18 22:16:47  a114682c4d0a1df6cafb53f9ce37af591cfd7bfb  SiPixel2DTemplateDBObject 
312209       2018-06-18 22:20:22  50758a12856b842ffe768ac29348d73bda9480b3  SiPixel2DTemplateDBObject 
324276       2018-10-10 19:02:50  916a68aa85d2bbbb4c61f0db8a1bdb6a30008829  SiPixel2DTemplateDBObject
```

an easy workaround would be if you could supply new Global Tag for the prompt-like workflows that instead of the tag `SiPixel2DTemplateDBObject_38T_v1_prompt` contain the tag: `SiPixel2DTemplateDBObject_38T_v1_offline`.
Please let us know.


[0]
```
$ conddb list 105X_dataRun2_PromptLike_v8 | grep SiPixel2D
SiPixel2DTemplateDBObjectRcd                            numerator                                         SiPixel2DTemplateDBObject_38T_v1_prompt                                        
$ conddb list 105X_dataRun2_PromptLike_HI_v3 | grep SiPixel2D
SiPixel2DTemplateDBObjectRcd                            numerator                                         SiPixel2DTemplateDBObject_38T_v1_prompt        
```

tvami commented:
Hi,
to make things a bit faster, I've queued `SiPixel2DTemplateDBObject_38T_v1_offline` to 
 `105X_dataRun2_PromptLike_Queue`
Tamas

tvami commented:
Hi @tocheng, I've created `SiPixel2DTemplateDBObject_38T_v2_prompt` as a copy of `SiPixel2DTemplateDBObject_38T_v1_offline` but with prompt synchronization and queued to the same queue as above.

tocheng committed:
solve conflicts with old local conditions

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9101

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9101/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9101/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

cmantill commented:
@mmusich @tocheng  I am confused by this patch:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9101/git-diff.patch
The last lines seem to update the GTs according to this PR: https://github.com/cms-sw/cmssw/commit/f7c4a8a05cf44727cafbd4b77286dab98b1985a0#diff-8cbfe6480e6018966eac475523946acd
which is a reasonable thing to fix.

However, the first lines are reverting the commit we just did to update pixel local reco conditions for 2017.

Please let me know how to proceed.

tocheng commented:
> @mmusich @tocheng I am confused by this patch:
> https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9101/git-diff.patch
> The last lines seem to update the GTs according to this PR: [f7c4a8a#diff-8cbfe6480e6018966eac475523946acd](https://github.com/cms-sw/cmssw/commit/f7c4a8a05cf44727cafbd4b77286dab98b1985a0#diff-8cbfe6480e6018966eac475523946acd)
> which is a reasonable thing to fix.
> 
> However, the first lines are reverting the commit we just did to update pixel local reco conditions for 2017.
> 
> Please let me know how to proceed.

@cmantill Can you rebase the PR?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9103

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

mmusich commented:
@perrotta FYI: we still expect few failures as the question of how to treat the prompt-like workflows is not resolved. Hopefully this PR: https://github.com/cms-sw/cmssw/pull/26281 is going to solve few of our problems, but not all...

perrotta commented:
Thank you @mmusich for telling
Let run the tests anyhow in the meanwhile and check if they witness some other possible issue

cmsbuild commented:
-1

Tested at: 4eb37bd66e5bef08629ab1c46bab3e2c31115f9d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/33993/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
101.0 step1
<pre>runTheMatrix-results/101.0_SingleElectronE120EHCAL+SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL+SingleElectronE120EHCAL.log

</pre>140.56 step2
<pre>runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log

</pre>136.788 step3
<pre>runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

</pre>136.85 step3
<pre>runTheMatrix-results/136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM/step3_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_PRef_DATA.root --fileout file:RelVal_Raw_PRef_DATA_HLT_RECO.root : FAILED - time: date Fri Apr  5 14:21:47 2019-date Fri Apr  5 14:15:48 2019 s - exit: 16640
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_PIon_DATA.root --fileout file:RelVal_Raw_PIon_DATA_HLT_RECO.root : FAILED - time: date Fri Apr  5 14:21:37 2019-date Fri Apr  5 14:16:09 2019 s - exit: 16640
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Fri Apr  5 14:26:56 2019-date Fri Apr  5 14:16:31 2019 s - exit: 16640

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

mmusich commented:
So, the errors for  `140.56 step2 `, `136.788 step3 `, `136.85 step3` are expected. 
Errors for 
  * `136.788` and `136.85` should go away once https://github.com/cms-sw/cmssw/pull/26281 is merged.
  * `140.56 step2 ` will need special customization in absence of updating the payload in the Prompt GT or allowing 1D vs 2D lookup ID mismatches in the code.

as for:
   *  `101.0 step1`

I am not sure at all this is due to this PR:

```
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/nweek-02570/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_6_X_2019-04-04-1100/bin/slc7_amd64_gcc700/cmsDriver.py", line 56, in <module>
    run()
  File "/cvmfs/cms-ib.cern.ch/nweek-02570/slc7_amd64_gcc700/cms/cmssw/CMSSW_10_6_X_2019-04-04-1100/bin/slc7_amd64_gcc700/cmsDriver.py", line 28, in run
    configBuilder.prepare()
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc700/cms/cmssw-patch/CMSSW_10_6_X_2019-04-04-2300/python/Configuration/Applications/ConfigBuilder.py", line 2251, in prepare
    self.pythonCfgCode += self.addCustomise()
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc700/cms/cmssw-patch/CMSSW_10_6_X_2019-04-04-2300/python/Configuration/Applications/ConfigBuilder.py", line 860, in addCustomise
    raise Exception("cannot specify twice "+fcn+" as a customisation method") 
Exception: cannot specify twice customise as a customisation method
```
looks like the `ConfigBuilder` is broken?

smuzaffar commented:
please test
cms-bot PR https://github.com/cms-sw/cms-bot/pull/1115 should fix 101

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 4eb37bd66e5bef08629ab1c46bab3e2c31115f9d

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34027/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
136.788 step3
<pre>runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log

</pre>140.56 step2
<pre>runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log

</pre>136.85 step3
<pre>runTheMatrix-results/136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM/step3_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_PRef_DATA.root --fileout file:RelVal_Raw_PRef_DATA_HLT_RECO.root : FAILED - time: date Sat Apr  6 11:32:38 2019-date Sat Apr  6 11:24:25 2019 s - exit: 16640
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_PIon_DATA.root --fileout file:RelVal_Raw_PIon_DATA_HLT_RECO.root : FAILED - time: date Sat Apr  6 11:32:34 2019-date Sat Apr  6 11:24:41 2019 s - exit: 16640
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2018 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Sat Apr  6 11:37:27 2019-date Sat Apr  6 11:25:14 2019 s - exit: 16640

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

OzAmram committed:
Allow different template IDs for 1D and 2D. Lookup all IDs during initialization

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9124

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

pohsun commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f346656bbb1fd186260846b6ddcabc0abcb912ef
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34038/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34038/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 10631 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 11429
 - DQMHistoTests: Total nulls: 7
 - DQMHistoTests: Total successes: 3128862
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.028 KiB( 31 files compared)
 - DQMHistoSizes: changed ( 136.85 ): -0.016 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 136.788 ): -0.012 KiB       JetMET/SUSYDQM
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

mmusich commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -91,6 +94,55 @@ PixelCPEClusterRepair::PixelCPEClusterRepair(edm::ParameterSet const & conf,
+
+      p.detTemplateId2D = templateDBobject2D_->getTemplateID(p.theDet->geographicalId());
+      if(p.detTemplateId != p.detTemplateId2D && !printed_info){
+          edm::LogInfo("PixelCPEClusterRepair") << "different template ID between 1D and 2D " << p.detTemplateId << " " << p.detTemplateId2D << endl;
```
@OzAmram @cmantill 
I thought that after your last commit we would have found a message about the mismatch in the workflow steps which were failing before, but I cannot find any in:
   * [136.788 step3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34038/runTheMatrix-results/136.788_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017/step3_RunSinglePh2017B+RunSinglePh2017B+HLTDR2_2017+RECODR2_2017reHLT_skimSinglePh_Prompt+HARVEST2017.log)
   * [140.56 step2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34038/runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18.log)
   * [136.85 step3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34038/runTheMatrix-results/136.85_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM/step3_RunEGamma2018A+RunEGamma2018A+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Prompt_L1TEgDQM+HARVEST2018_L1TEgDQM.log)

shall we promote this to `LogWarning` instead?

OzAmram replied:
Hi Marco. Yes I think if we want them to be printed out we should promote it to LogWarning. When I tested locally, just running a runTheMatrix workflow did not print out the message, but adding a messageLogger to the config file did produce the message.

I will test right now that upgrading to LogWarning does print out a message for standard workflows

OzAmram replied:
Yes I can confirm that upgrading to LogWarning does print out a message in the standard workflow. So I suggest we do that

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -91,6 +94,55 @@ PixelCPEClusterRepair::PixelCPEClusterRepair(edm::ParameterSet const & conf,
+   //     XYZ n (XYZ as above, n = layer, wheel or disk = 1 .. 6 ;)
+   std::vector<std::string> str_Recommend2D = conf.getParameter<std::vector<std::string> >("Recommend2D");
+   Recommend2D_.reserve(str_Recommend2D.size());
+   for (std::vector<std::string>::const_iterator it = str_Recommend2D.begin(), ed = str_Recommend2D.end(); it != ed; ++it) {
```
Range based for loop must be preferred here

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -91,6 +94,55 @@ PixelCPEClusterRepair::PixelCPEClusterRepair(edm::ParameterSet const & conf,
+   }
+
+   // run CR on damaged clusters (and not only on edge hits)
+   RunDamagedClusters_ = conf.existsAs<bool>("RunDamagedClusters")?conf.getParameter<bool>("RunDamagedClusters"):false;
```
Should you have defined a fillDescriptions() method in this class. this check wouldn't have been needed/
Defining such a method here shouldn't be complicated at all: however, if you don't want to do it now please keep it in mind for the post UL improvements/updates

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -492,10 +544,15 @@ void PixelCPEClusterRepair::checkRecommend2D( DetParam const & theDetParam, Clus
-        //only run on barrel
+
+    bool recommend = false;
+    for (std::vector<Rule>::const_iterator itr = Recommend2D_.begin(), edr = Recommend2D_.end(); itr != edr; ++itr) {
```
Range based for loop must be preferred here

perrotta commented on RecoLocalTracker/SiPixelRecHits/interface/PixelCPEClusterRepair.h:
```diff
@@ -104,13 +106,40 @@ class PixelCPEClusterRepair : public PixelCPEBase
+     int  subdet_;
+     int  layer_;
+   };
+   std::vector<Rule> Recommend2D_;
```
Please start local variables with lowercase letters, as per CMS code rules

perrotta commented on RecoLocalTracker/SiPixelRecHits/interface/PixelCPEClusterRepair.h:
```diff
@@ -104,13 +106,40 @@ class PixelCPEClusterRepair : public PixelCPEBase
+   std::vector<Rule> Recommend2D_;
+
+   // run on damaged hits or not
+   bool RunDamagedClusters_;
```
Please start local variables with lowercase letters, as per CMS code rules

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -130,18 +182,18 @@ PixelCPEClusterRepair::localPosition(DetParam const & theDetParam, ClusterParam
+	ID1 = ID2 = forwardTemplateID_ ; // forward
    }
-   //cout << "PixelCPEClusterRepair : ID = " << ID << endl;
+   //cout << "PixelCPEClusterRepair : ID1 = " << ID1 << endl;
```
Remove this commented out line

perrotta commented:
Some increase in cpu timing of approx 250 ms/evt was measured, before the integration of the last commit, on TTbar PU events (wf 11024):

```
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +0.402955      +0.10%        30.48 ms/ev ->        45.87 ms/ev mixedTripletStepTracks
   +0.302630      +0.01%         3.40 ms/ev ->         4.61 ms/ev jetCoreRegionalStepTracks
   +0.223082      +0.26%       164.41 ms/ev ->       205.69 ms/ev initialStepTracksPreSplitting
   +0.216801      +0.22%       141.32 ms/ev ->       175.68 ms/ev detachedTripletStepTracks
   +0.215750      +0.24%       156.37 ms/ev ->       194.18 ms/ev highPtTripletStepTracks
   +0.194436      +0.25%       185.60 ms/ev ->       225.57 ms/ev initialStepTracks
   +0.182334      +0.20%       160.32 ms/ev ->       192.48 ms/ev lowPtTripletStepTracks
   +0.170239      +0.21%       176.23 ms/ev ->       209.02 ms/ev lowPtQuadStepTracks
  ---------- ------------     --------                  ----       ------------
Job total:  15.9665 s/ev ==> 16.2681 s/ev 
```

The time increase due to the siPixelRecHits themselves was already measured to be rather negligible when the full cluster repair was admitted, see slides linked in the description of previous PR #25727, and even here it is below the resolution I've chosen for the measurement method.

On the other hand, the time increase in these tracking iterations is not negligible, and it probably contributed also to the overall 7% overall reco cpu time increase that was reported unexplained in those slides (here the reco+miniAOD CPU time increase is measured slightly below 2%).

Could you please remind us where that cpu time increase in the tracking iterations comes from?
Can you provide some dedicated measurement of performance?

cmantill committed:
switch to LogWarning instead

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9138

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

cmantill commented:
@perrotta I performed a timing test on data using a JetHT 2017 file [1] for 100 evts with the PR and without using the timing RECO tool [2] and obtained the following (attached [3]) result. For a better comparison, I put here the same modules you just showed (not sure how to run the timing comparison for certain modules only):

```
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +0.316399      +0.05%        21.63 ms/ev ->        29.76 ms/ev mixedTripletStepTracks
   +0.224855      +0.02%        10.75 ms/ev ->        13.47 ms/ev jetCoreRegionalStepTracks
   +0.182794      +0.14%       115.68 ms/ev ->       138.95 ms/ev initialStepTracksPreSplitting
   +0.142637      +0.12%       125.10 ms/ev ->       144.32 ms/ev detachedTripletStepTracks
   +0.140086      +0.13%       138.56 ms/ev ->       159.43 ms/ev highPtTripletStepTracks
   +0.148603      +0.13%       133.62 ms/ev ->       155.07 ms/ev initialStepTracks
   +0.122316      +0.11%       142.40 ms/ev ->       160.96 ms/ev lowPtTripletStepTracks
   +0.124279      +0.11%       132.33 ms/ev ->       149.86 ms/ev lowPtQuadStepTracks
  ---------- ------------     --------                  ----       ------------
Job total:  16.1396 s/ev ==> 15.3975 s/ev
```
I am confused though because the total #s/ev seems to be reduced. 

We think that the increase on time for tracking iterations could be due to an increase of # of reconstructed tracks (which we are not sure if its significant - this was seen in the old RelVals that had CPE running for damaged hits as well - but will check DQM comparison plots of the current PR now).

We expect better hit position estimates but CPE also makes the errors slightly larger which could result in an increase of the # tracks. However we expect these differences on the # of tracks to be very small (<1%) so we are not sure if this is causing the increase of tracking time shown in these modules. 
We will look at the DQM plots in detail now. I will also keep making the modifications requested in the PR.

[1] /store/data/Run2017B/JetHT/RAW/v1/000/297/618/00000/04A34499-EA5B-E711-A0EC-02163E01A4E6.root'
[2] https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#RECO_tools
[3] http://cmantill.web.cern.ch/cmantill/tmp/OFFnew_vs_ON.txt

perrotta commented:
@cmantill thank you for your check

For the timing, please check that there are no other processes which are running concurrently when you test it. At the first order, you can renormalize on some process that you know cannot be affected by this PR and make the comparison once again.

makortel commented:
The modules in question are the ones performing the final fit of the tracks. They are also the only modules within the tracking iterations making use of the pixel template CPE (also duplicate merging modules  seem to make use of templates). Could the time increase be then just the cost of the cluster repair?

(by the way, how can the cluster repair affect the timing of siPixelRecHits if the repair is not used in there?)

OzAmram commented:
@makortel I think you are correct that siPixelRecHits was the wrong module to analyze when I did the timing studies for the PR in January. I was confused about the module structure of what is actually calling ClusterRepair. I apologize for that error. 

Still I would be surprised if the percentage increases we are seeing are just due to the cluster repair itself. Its only running the 2D fit on around 2% of the hits and the 2D fit should only take a few times longer than the 1D fit. So I would expect a few % increase. But the differences we are seeing from those module is around 25-30% increase in time.

makortel commented:
Thanks @OzAmram. The `initialStepTracksPreSplitting` is the very first iteration for which the output of the pattern recognition should be exactly the same between the two cases (for `initialStepTracks` the jet-core pixel cluster splitting could have an effect), and in the reconstruction chain from the raw data that module should be first one to make use of template/repair CPE.

Maybe it would be worth to investigate the difference with e.g. igprof?

cmantill committed:
fix variable names and comments

cmantill commented:
@makortel do you have a particular command to be used for igprof? I tried `igprof -d -pp -t cmsRun -o igprof.pp cmsRun step3_igprof.py` and seems to be working at first but when I try to read the output with ` igprof-analyse -g -d -v -p -r MEM_MAX igprof.pp` I get no results, see [1].  Do you know if there are better commands to test it?

Also, is there a way to profile those particular tracking modules only?

[1] http://cmantill.web.cern.ch/cmantill/tmp/igprof.pp
http://cmantill.web.cern.ch/cmantill/tmp/igproflog

cmantill committed:
add fillDescriptions method for CPE CR

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9148

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

perrotta commented:
> @makortel do you have a particular command to be used for igprof? I tried `igprof -d -pp -t cmsRun -o igprof.pp cmsRun step3_igprof.py` and seems to be working at first but when I try to read the output with ` igprof-analyse -g -d -v -p -r MEM_MAX igprof.pp` I get no results, see [1]. Do you know if there are better commands to test it?
> 

@cmantill You can find instructions in https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Run_profiler_igprof

At a first glance, you should analyse for cpu, non for MEM_MAX. Following those instructions:
```
igprof-analyse --demangle --gdb igprof.pp
```

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -94,18 +94,18 @@ PixelCPEClusterRepair::PixelCPEClusterRepair(edm::ParameterSet const & conf,
-     Recommend2D_.push_back(Rule(*it));
+   std::vector<std::string> str_recommend2D = conf.getParameter<std::vector<std::string> >("Recommend2D");
+   recommend2D_.reserve(str_recommend2D.size());
+   for (unsigned int i=0; i<str_recommend2D.size();++i){
```
What I had in mind was something as
```C++
   for (auto & str: str_recommend2D){
     recommend2D_.push_back(str);
   }
```

perrotta commented on RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:
```diff
@@ -546,8 +545,8 @@ void PixelCPEClusterRepair::checkRecommend2D( DetParam const & theDetParam, Clus
     bool recommend = false;
-    for (std::vector<Rule>::const_iterator itr = Recommend2D_.begin(), edr = Recommend2D_.end(); itr != edr; ++itr) {
-        recommend = itr->recommend(id, ttopo_);
+    for (unsigned int i=0; i<recommend2D_.size(); i++){
```
What I had in mind was something as
```C++
   for (auto & rec: recommend2D_){
        recommend = rec.recommend(id, ttopo_);
	if(recommend) break;
    }
```

cmsbuild commented:
+1
Tested at: 7df3dad1743211840401209c6194c80dc04fc934
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34064/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34064/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 10612 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 11292
 - DQMHistoTests: Total nulls: 7
 - DQMHistoTests: Total successes: 3128999
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.012 KiB( 31 files compared)
 - DQMHistoSizes: changed ( 136.85 ): -0.020 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 136.788 ): 0.008 KiB        JetMET/SUSYDQM
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

cmantill committed:
range based for loop

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9161

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

cmantill commented:
@perrotta sorry for the delay I have been getting a lot of errors either when running igprof or when trying to convert to sqlite file.

Here is the full igprof report for 10 evts in JetHT with CRon (if I try more the script crashes even though it runs fine without igprof):
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig

A couple of interesting notes are the following:
- By cumulative-cost, we think that the performance of `PixelCPEClusterRepair::localPosition` could explain the increased CPU time in the tracking modules:
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig/330
If you continue to follow you can see that the functions callTempReco1D and 2D are the most time consuming and that 2D reco takes about 35% of the time. It seems that is actually the fit that is taking most of the CPU time:
https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig/469

This would explain the differences seen in CPU time. Because we are only running the 2D on edge clusters (about 2% of hits), this would imply that the 2D fit takes 15 to 20 times longer than the 1D fit. This is a lot slower than we had expected so we will talk to Morris to see if he thinks that makes sense.

I also have the following txt comparison for CRon/CROff:
http://cmantill.web.cern.ch/cmantill/tmp/CRON_igprof_10evts.res
http://cmantill.web.cern.ch/cmantill/tmp/CROFF_igprof_10evts.res

For some reason I can't get the sqlite file to work. But since the .txt files are too big (~20MB) I grep the modules where PixelTemplateReco is called:
http://cmantill.web.cern.ch/cmantill/tmp/CRON_igprof_10evts_ClusterRepair.res
http://cmantill.web.cern.ch/cmantill/tmp/CROFF_igprof_10evts_ClusterRepair.res
I will try to get the sqlite version working in the meantime.

cmantill committed:
set recommend2D to false by default

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26263/9210

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #26263 was updated. @perrotta, @christopheralanwest, @tocheng, @cmsbuild, @franzoni, @slava77, @pohsun can you please check and sign again.

cmantill commented:
@perrotta We found a bug in the current implementation to call the 2D fit: 

For cases where we are in the correct layers (BPIX 2,3,4) but not edgehits, the theClusterParam.recommended2D_ is set to a garbage value (!=0) that later passes the 2D call. Thus, instead of calling 2D reco for 0.5% of the hits, we were calling for 10% of them. We have fixed that in the latest commit by setting theClusterParam.recommended2D_  false by default.

I should clarify that this bug should have not affected the hit residuals by a significant amount. For the hits affected by this bug, the 2D Reco would have been running with option edgeY=0, i.e. when CR is not trying to repair anything and is pretty much the same as the 1D Reco. It was only making things much slower as shown by the results below.

With the fix introduced in the latest commit we have updated the timing and CPU time tests as follows:
- Here is the CPU time comparison with CRON, again running on JetHT for 10evts (withbugfix), https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/igprofbugpixCRON/342  compared to:
  - CROFF (https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/igprofCROFF10evts/394)
  - CRON with bug (https://cmantill.web.cern.ch/cmantill/cgi-bin//igprof-navigator/ig/330)
There is a significant reduction of the CPU time that 2D consumes: 2D uses 5% of the CPU time, while the 1D fit consumes most of it ~90%. 

- Here [1] is the timing comparison run on the same 100 evts of JetHT as before (in lxplus too) compared to the old log with CROFF: 
```   +0.098533      +0.01%        20.75 ms/ev ->        22.91 ms/ev mixedTripletStepTracks
   +0.062444      +0.00%        10.43 ms/ev ->        11.10 ms/ev jetCoreRegionalStepTracks
   +0.057465      +0.04%       113.48 ms/ev ->       120.19 ms/ev initialStepTracksPreSplitting
   +0.065809      +0.05%       121.00 ms/ev ->       129.24 ms/ev detachedTripletStepTracks
   +0.070813      +0.06%       134.83 ms/ev ->       144.73 ms/ev highPtTripletStepTracks
   +0.071519      +0.06%       130.54 ms/ev ->       140.22 ms/ev initialStepTracks
   +0.069876      +0.06%       138.94 ms/ev ->       149.00 ms/ev lowPtTripletStepTracks
   +0.087252      +0.07%       128.88 ms/ev ->       140.64 ms/ev lowPtQuadStepTracks
Job total:  15.7222 s/ev ==> 17.0266 s/ev
```
Although the time increases for tracking, the increased time is less than that shown before the bugfix. We think that the overall increase of time is due to random increases of other modules (also this comparison is done to a test with CROFF two days ago).

[1] https://cmantill.web.cern.ch/cmantill/tmp/newtiming_afterbugfix.log

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: bec1c249ee7a17f4e150c1e79a92c30122b30f48
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34137/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26263/34137/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8476 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 12869
 - DQMHistoTests: Total nulls: 6
 - DQMHistoTests: Total successes: 3127423
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.028 KiB( 31 files compared)
 - DQMHistoSizes: changed ( 136.788 ): -0.016 KiB       JetMET/SUSYDQM
 - DQMHistoSizes: changed ( 136.85 ): -0.012 KiB       JetMET/SUSYDQM
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

perrotta commented:
After the last fix, overall timing is quite comparable while using ClusterRepair on a limited number of hits, as in this PR, or having it turned off, as in the baseline IB. I think that the differences you report in https://github.com/cms-sw/cmssw/pull/26263#issuecomment-481808539 are mostly due to an overall normalization factor, as I see no relevant increase while running on the usual 100 evt of the TTbar PU sample 11024.

The number of hits affected being now rather limited, as you pointed out, also the effects on general tracks justshow up as small deviations from the baseline, see
![image](https://user-images.githubusercontent.com/4069749/55939503-30533d00-5c3e-11e9-9d9f-329049224e6f.png)
The same for the  derived higher order objects, see for example 
![image](https://user-images.githubusercontent.com/4069749/55939719-c0918200-5c3e-11e9-9b9f-1b65dc9ed712.png)
or
![image](https://user-images.githubusercontent.com/4069749/55939799-fafb1f00-5c3e-11e9-8270-885a19e056e9.png)

A detailed validation cannot be done with just those small test samples, but some is certainly expected based on pre4, provided this PR can get merged for it

perrotta commented:
+1
- ClusterRepair has beed turned on for the restricted class of hits as in the PR descritption.
- While restricting to that set of hits the impact on code performance is negligible, and also the propagation to the reco objects has overall a limited effect.
- Jenkins tests pass and show small deviations in tracking and derived quantities, as expected.

pohsun commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
